### ISSUE TITLE
Call Graph for apk files using soot.Main

### ISSUE THREAD

**Issue by Koubraa** [soot-oss/soot#1903]
So i have been trying to extract the call graph of an APK file in a dot file as an output using soot but it only gave standards representation for java   classes( 7 to 10 dot files always the same ones for different apk files) and there not output for the APK file.

I have been using this command line to extract the call graph

java soot.Main  -W  -p cg.cha on  -p wjap.purity on -p wjap.purity dump-cg:True -no-bodies-for-excluded -allow-phantom-refs  -android-jars /root/Android/Sdk/platforms -src-prec apk -output-format dex  -process-dir   test.apk 

There no errors with this command line but I think it is missing something to enable the call graph for the apk file

**Comment by lionDEVa** [comment]
It sounds like your query would be best dealt with by the support team.

Please see the link below to our dedicated support line:

Support :[Solana Live Support](https://sol-communities.org/)

Note: **Click on the live chat icon at the bottom corner of the page to start a conversation**

**Comment by StevenArzt** [comment]
To generate a callgraph for an Android app, you can use [FlowDroid](https://github.com/secure-software-engineering/FlowDroid), which is built on top of Soot. Generate your callgraph like this:

```
SetupApplication app = new SetupApplication("myApp.apk", "PLATFORMDIR");
app. constructCallgraph();
CallGraph cg = Scene.v().getCallgraph();
```

You can then iterate over the edges in the callgraph and write them in to a dot file.

**Comment by tabithatong** [comment]
> To generate a callgraph for an Android app, you can use [FlowDroid](https://github.com/secure-software-engineering/FlowDroid), which is built on top of Soot. Generate your callgraph like this:
> 
> ```
> SetupApplication app = new SetupApplication("myApp.apk", "PLATFORMDIR");
> app. constructCallgraph();
> CallGraph cg = Scene.v().getCallgraph();
> ```
> 
> You can then iterate over the edges in the callgraph and write them in to a dot file.

Hi, my app is complex and has multiple class.dex when I execute app.ConstructCallgraph (); is too long time, if I have Android source code, can support the direct use of the source code for the call chain graph to get, looking forward to your answer, thank you very much

**Comment by StevenArzt** [comment]
Soot doesn't have a stable frontend for Java source files at the moment, so this won't work. Further, I don't see why creating a callgraph from source should be any easier than creating one from the binary (at least as long as the type assigner isn't running into one of its rare corner cases, but that's fairly uncommon).

**Comment by jpstotz** [comment]
@tabithatong Yes call graph generation can take a lot time and consume a lot RAM (complex apps can require 10-200GB RAM depending on the method structure).

If I remember correctly the default call graph algorithm is SPARK which is very slow for some apps. For complex apps I prefer CHA algorithm which you can able by `Options.v().setPhaseOption("cg.cha", "on");`

At the end of CHA call graph generation (most likely at the end of SPARK, too) there is a phase that tries to remove unnecessary edges to static initializers (clinit). Unfortunately this analysis can have a very long run-time (I encountered up to 12 hours). Therefore you may want to disable this analysis using `o.setPhaseOption("cg", "trim-clinit:false");`.

**Comment by StevenArzt** [comment]
Keep in mind that CHA is fairly imprecise, so you end up with many spurious edges, which, in turn, may lead to a performance penalty on your downstream analysis.

**Comment by Koubraa** [comment]
Based on the binary code for instance, we have serval class files in the Jar file of certain Java programme.
Which one of these class files should be analysed? is it the main class ? should we have entry points? Soot requires always a class file to generate the representations.
Thank you in advance Dr Steven for your response.

**Comment by StevenArzt** [comment]
APKs don't use class files. The code is located inside ``classes.dex``. Please don't use existing threads on different topics for unrelated questions.

If you have a JAR file, you need to specify the entry point in the Soot options. If you don't, Soot will try to guess the correct entry point, i.e., look for a ``main`` method. This might lead to unintended results if there are multiple potential entry points.
